src/iconforge.rs

```diff
@@ -0,0 +1,551 @@
+// DMI spritesheet generator
+// Developed by itsmeow
+use crate::jobs;
+use crate::hash::string_hash;
+use crate::error::Error;
+use std::{
+    fs::File,
+    io::BufReader,
+    sync::{Arc, Mutex},
+    collections::HashMap,
+};
+use dmi::icon::{Icon, IconState};
+use image::{DynamicImage, GenericImage, GenericImageView, Pixel, ImageBuffer};
+use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
+//use raster::Image;
+use serde::{Serialize, Deserialize};
+use dashmap::DashMap;
+use tracy_full::{zone, frame};
+use once_cell::sync::Lazy;
+static ICON_FILES: Lazy<DashMap<String, Arc<Icon>>> = Lazy::new(DashMap::new);
+static ICON_STATES: Lazy<DashMap<String, DynamicImage>> = Lazy::new(DashMap::new);
+
+const SOUTH: u8 = 2;
+const NORTH: u8 = 1;
+const EAST: u8 = 4;
+const WEST: u8 = 8;
+const FOUR_DIRS: [u8; 4] = [SOUTH, NORTH, EAST, WEST];
+const SOUTHEAST: u8 = SOUTH | EAST; // 6
+const SOUTHWEST: u8 = SOUTH | WEST; // 10
+const NORTHEAST: u8 = NORTH | EAST; // 5
+const NORTHWEST: u8 = NORTH | WEST; // 9
+// This is ordered by how DMIs internally place dirs into the PNG
+const EIGHT_DIRS: [u8; 8] = [SOUTH, NORTH, EAST, WEST, SOUTHEAST, SOUTHWEST, NORTHEAST, NORTHWEST];
```

Comment:
Probably should be in some shared module or it's own lib tbh

Reply:
It should really be part of the DMI crate, heh